packages/@lwc/engine-core/src/framework/hydration.ts

```diff
@@ -505,11 +505,12 @@ function validateClassAttr(vnode: VBaseElement, elm: Element, renderer: Renderer
     let nodesAreCompatible = true;
     let readableVnodeClassname;
 
-    const elmClassName = getProperty(elm, 'className');
-    if (!isUndefined(className) && String(className) !== elmClassName) {
+    const elmClassName = getAttribute(elm, 'class');
```

Comment:
I realized the main problem is using `elm.className` rather than `elm.getAttribute('class')`. The latter gives more consistent results than the former (since SVGs behave differently than HTML).